src/Microsoft.AspNetCore.SignalR.Client/RpcConnection.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Sockets.Client;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.SignalR.Client
+{
+    public class RpcConnection : IDisposable
+    {
+        private readonly IInvocationAdapter _adapter;
+        private readonly Connection _connection;
+        private readonly Stream _stream;
+        private readonly Task _reader;
+
+        private readonly CancellationTokenSource _readerCts = new CancellationTokenSource();
+        private readonly ConcurrentDictionary<string, InvocationRequest> _pendingCalls = new ConcurrentDictionary<string, InvocationRequest>();
+
+        private int _nextId = 0;
+
+        private RpcConnection(Connection connection, IInvocationAdapter adapter)
```

Comment:
Agreed, we may want an interface. Ideally we could just use a Pipeline, but I'm not sure if that fits in well with the necessary lifetime events... though we could probably model those with Complete on the Pipeline.